rfcs/0042-local-test-framework.md

```diff
@@ -0,0 +1,272 @@
+
+These logs can be polled on an interval and processed by a filter as they come in. 
+
+## Interface To Develop:
+
+The current logging for the cloud framework is done by creating a Google Stackdriver subscription and issuing poll requests for logs while doing some pre-defined filtering. 
+
+An example of this is shown below:
+
+```ocaml
+let rec pull_subscription_in_background ~logger ~network ~event_writer
+    ~subscription =
+  if not (Pipe.is_closed event_writer) then (
+    [%log spam] "Pulling StackDriver subscription" ;
+    let%bind log_entries =
+      Deferred.map (Subscription.pull ~logger subscription) ~f:Or_error.ok_exn
+    in
+    if List.length log_entries > 0 then
+      [%log spam] "Parsing events from $n logs"
+        ~metadata:[("n", `Int (List.length log_entries))]
+    else [%log spam] "No logs were pulled" ;
+    let%bind () =
+      Deferred.List.iter ~how:`Sequential log_entries ~f:(fun log_entry ->
+          log_entry
+          |> parse_event_from_log_entry ~network
+          |> Or_error.ok_exn
+          |> Pipe.write_without_pushback_if_open event_writer ;
+          Deferred.unit )
+    in
+    let%bind () = after (Time.Span.of_ms 10000.0) in
+    pull_subscription_in_background ~logger ~network ~event_writer
+      ~subscription )
+  else Deferred.unit
+```
+
+[https://github.com/MinaProtocol/mina/blob/67cc4205cc95138cf729a2f14b57b754f9e9204e/src/lib/integration_test_cloud_engine/stack_driver_log_engine.ml#L269](https://github.com/MinaProtocol/mina/blob/67cc4205cc95138cf729a2f14b57b754f9e9204e/src/lib/integration_test_cloud_engine/stack_driver_log_engine.ml#L269)
+
+A similar interface can be written for Docker-Swarm instead. By defining a `Service.pull` function with a given logger, we can leverage a lot of the work already done by modifying parts of the code where the log formats diverge. All logs can be specified to an output stream, such as stdout or a specified file by the user on their local system. 
+
+<img src="./res/local-test-integration-logging.png" alt="drawing" width="500"/>
+
+# Work Breakdown/Prio
+
+The following will be a work breakdown of what needs to be done to see this feature to completion:
+
+1. Implement the `Network_Config` interface to accept a network configuration and create a corresponding `docker-compose.json` file.
+2. Implement the `Network_manager` interface to take a corresponding `docker-compose.json` file and create a local swarm with the specified container configuration
+3. Implement functionality to simply log all container logs into a single stream (stdout or a file, maybe this can be specified in startup?)
+4. Implement filter on event functionality
+5. Ensure that current integration test specs are able to run on the local framework with success
+
+
+# Unresolved Questions
+
+- Is compiling a docker-compose file the right approach for scheduling the containers? The nice thing about using a docker-compose file is that all network management should be automatic.
```

Comment:
Yeah, I think this is the right way to go, especially for the network management bit.